src/main/java/it/gov/pagopa/payhub/ionotification/exception/IONotificationExceptionHandler.java

```diff
@@ -0,0 +1,101 @@
+import org.springframework.core.Ordered;
+import org.springframework.core.annotation.Order;
+import org.springframework.dao.DataAccessException;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@RestControllerAdvice
+@Slf4j
+@Order(Ordered.HIGHEST_PRECEDENCE)
+public class IONotificationExceptionHandler {
+
+    private static final Pattern RETRY_AFTER_MS_PATTERN = Pattern.compile("RetryAfterMs=(\\d+)");
+
+    @ExceptionHandler({RetrieveServicesInvocationException.class, CreateServiceInvocationException.class})
+    public ResponseEntity<IoNotificationErrorDTO> handleFeignClientException(RuntimeException ex, HttpServletRequest request) {
+        return handleIONotificationErrorException(ex, request, HttpStatus.INTERNAL_SERVER_ERROR, IoNotificationErrorDTO.CodeEnum.GENERIC_ERROR);
+    }
+
+    @ExceptionHandler({IOWrongPayloadExceptiion.class})
+    public ResponseEntity<IoNotificationErrorDTO> handleWrongPayloadException(RuntimeException ex, HttpServletRequest request) {
+        return handleIONotificationErrorException(ex, request, HttpStatus.BAD_REQUEST, IoNotificationErrorDTO.CodeEnum.WRONG_PAYLOAD);
+    }
+
+    @ExceptionHandler(DataAccessException.class)
+    protected ResponseEntity<IoNotificationErrorDTO> handleDataAccessException(
+            DataAccessException ex, HttpServletRequest request) {
+
+        if (isRequestRateTooLargeException(ex)) {
+            Long retryAfterMs = getRetryAfterMs(ex);
+            return getErrorDTOResponseEntity(ex, request, retryAfterMs);
+        } else {
+            return handleIONotificationErrorException(ex, request, HttpStatus.INTERNAL_SERVER_ERROR, IoNotificationErrorDTO.CodeEnum.GENERIC_ERROR);
+        }
+    }
+
+    private ResponseEntity<IoNotificationErrorDTO> handleIONotificationErrorException(
+            RuntimeException ex, HttpServletRequest request, HttpStatus httpStatus, IoNotificationErrorDTO.CodeEnum errorEnum){
+        String message = ex.getMessage();
+        log.info("A {} occurred handling request {}: HttpStatus {} - {}",
+                ex.getClass(),
+                getRequestDetails(request),
+                httpStatus.value(),
+                message);
+
+        return ResponseEntity
+                .status(httpStatus)
+                .body(new IoNotificationErrorDTO(errorEnum, message));
+    }
+
+    private String getRequestDetails(HttpServletRequest request) {
+        return "%s %s".formatted(request.getMethod(), request.getRequestURI());
+    }
+
+    private ResponseEntity<IoNotificationErrorDTO> getErrorDTOResponseEntity(Exception ex, HttpServletRequest request, Long retryAfterMs) {
```

Comment:
let's rename this method... its purpose is to handle the cosmos 429 error, or requestRateTooLarge error..